extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/HybridProtobufReader.java

```diff
@@ -0,0 +1,141 @@
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+
+import static org.apache.druid.data.input.opencensus.protobuf.HybridProtobufReader.ProtobufReader.OPEN_CENSUS;
+import static org.apache.druid.data.input.opencensus.protobuf.HybridProtobufReader.ProtobufReader.OPEN_TELEMETRY;
+
+public class HybridProtobufReader implements InputEntityReader
+{
+  private static final String VERSION_HEADER_KEY = "v";
+  private static final int OPENTELEMETRY_FORMAT_VERSION = 1;
+
+  private final DimensionsSpec dimensionsSpec;
+  private final SettableByteEntity<? extends ByteEntity> source;
+  private final String metricDimension;
+  private final String valueDimension;
+  private final String metricLabelPrefix;
+  private final String resourceLabelPrefix;
+
+  private volatile MethodHandle getHeaderMethod = null;
+
+  enum ProtobufReader {
+    OPEN_CENSUS,
+    OPEN_TELEMETRY
+  }
+
+  public HybridProtobufReader(
+      DimensionsSpec dimensionsSpec,
+      SettableByteEntity<? extends ByteEntity> source,
+      String metricDimension,
+      String valueDimension,
+      String metricLabelPrefix,
+      String resourceLabelPrefix
+  )
+  {
+    this.dimensionsSpec = dimensionsSpec;
+    this.source = source;
+    this.metricDimension = metricDimension;
+    this.valueDimension = valueDimension;
+    this.metricLabelPrefix = metricLabelPrefix;
+    this.resourceLabelPrefix = resourceLabelPrefix;
+  }
+
+  @Override
+  public CloseableIterator<InputRow> read() throws IOException
+  {
+    return newReader(whichReader()).read();
+  }
+
+  public InputEntityReader newReader(ProtobufReader which)
+  {
+    switch(which) {
+      case OPEN_TELEMETRY:
+        return new OpenTelemetryMetricsProtobufReader(
+            dimensionsSpec,
+            source,
+            metricDimension,
+            valueDimension,
+            metricLabelPrefix,
+            resourceLabelPrefix
+        );
+      case OPEN_CENSUS:
+      default:
+        return new OpenCensusProtobufReader(
+            dimensionsSpec,
+            source,
+            metricDimension,
+            metricLabelPrefix,
+            resourceLabelPrefix
+        );
+    }
+  }
+
+  public ProtobufReader whichReader()
+  {
+    // assume InputEntity is always defined in a single classloader (the kafka-indexing-service classloader)
+    // so we only have to look it up once. To be completely correct we should cache the method based on classloader
+    if (getHeaderMethod == null) {
+      getHeaderMethod = KafkaUtils.lookupGetHeaderMethod(
+          source.getEntity().getClass().getClassLoader(),
+          VERSION_HEADER_KEY
+      );
+    }
+
+    try {
+      byte[] versionHeader = (byte[]) getHeaderMethod.invoke(source.getEntity());
+      if (versionHeader != null) {
+        int version =
+            ByteBuffer.wrap(versionHeader).order(ByteOrder.LITTLE_ENDIAN).getInt();
+        if (version == OPENTELEMETRY_FORMAT_VERSION) {
+          return OPEN_TELEMETRY;
```

Comment:
nit, for consistency we might want to keep OPENTELEMETRY as one word in constants (same with OPENCENSUS)

Reply:
Makes total sense.